File: packages/core/web/tests/retry.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { retry } from "../src";

describe("retry", () => {
  it("returns the value on first success", async () => {
    const result = await retry(() => Promise.resolve(42));
    expect(result).toBe(42);
  });

  it("works with synchronous functions", async () => {
    const result = await retry(() => "hello");
    expect(result).toBe("hello");
  });

  it("retries on failure and succeeds on subsequent attempt", async () => {
    let attempt = 0;
    const result = await retry(
      () => {
        attempt++;
        if (attempt < 3) return Promise.reject(new Error(`fail ${attempt}`));
        return Promise.resolve("ok");
      },
      { maxAttempts: 3, baseDelay: 1 },
    );

    expect(result).toBe("ok");
    expect(attempt).toBe(3);
  });

  it("throws the last error after all attempts are exhausted", async () => {
    await expect(
      retry(() => Promise.reject(new Error("always fails")), {
        maxAttempts: 2,
        baseDelay: 1,
      }),
    ).rejects.toThrow("always fails");
  });

  it("defaults to 3 maxAttempts", async () => {
    let attempt = 0;
    await expect(
      retry(
        () => {
          attempt++;
          return Promise.reject(new Error("fail"));
        },
        { baseDelay: 1 },
      ),
    ).rejects.toThrow("fail");
    expect(attempt).toBe(3);
  });

  describe("exponential backoff timing", () => {
    beforeEach(() => {
      vi.useFakeTimers();
    });

    afterEach(() => {
      vi.useRealTimers();
    });

    it("applies exponential backoff", async () => {
      let attempt = 0;
      const fn = () => {
        attempt++;
        if (attempt <= 3) return Promise.reject(new Error("fail"));
        return Promise.resolve("done");
      };

      const promise = retry(fn, {
        maxAttempts: 4,
        baseDelay: 100,
        multiplier: 2,
      });

      // attempt 1 fails → delay 100ms (100 * 2^0)
      await vi.advanceTimersByTimeAsync(100);
      expect(attempt).toBe(2);

      // attempt 2 fails → delay 200ms (100 * 2^1)
      await vi.advanceTimersByTimeAsync(200);
      expect(attempt).toBe(3);

      // attempt 3 fails → delay 400ms (100 * 2^2)
      await vi.advanceTimersByTimeAsync(400);
      const result = await promise;
      expect(result).toBe("done");
      expect(attempt).toBe(4);
    });

    it("caps delay at maxDelay", async () => {
      let attempt = 0;
      const fn = () => {
        attempt++;
        if (attempt <= 3) return Promise.reject(new Error("fail"));
        return Promise.resolve("done");
      };

      const promise = retry(fn, {
        maxAttempts: 4,
        baseDelay: 100,
        multiplier: 10,
        maxDelay: 500,
      });

      // attempt 1 fails → delay min(100 * 10^0, 500) = 100
      await vi.advanceTimersByTimeAsync(100);
      // attempt 2 fails → delay min(100 * 10^1, 500) = 500
      await vi.advanceTimersByTimeAsync(500);
      // attempt 3 fails → delay min(100 * 10^2, 500) = 500
      await vi.advanceTimersByTimeAsync(500);

      const result = await promise;
      expect(result).toBe("done");
    });
  });

  it("retryIf: retries when predicate returns true", async () => {
    let attempt = 0;
    const result = await retry(
      () => {
        attempt++;
        if (attempt < 2) return Promise.reject(new Error("transient"));
        return Promise.resolve("ok");
      },
      {
        baseDelay: 1,
        retryIf: (err) => (err as Error).message === "transient",
      },
    );
    expect(result).toBe("ok");
  });

  it("retryIf: rethrows immediately when predicate returns false", async () => {
    let attempt = 0;
    await expect(
      retry(
        () => {
          attempt++;
          return Promise.reject(new Error("fatal"));
        },
        {
          maxAttempts: 5,
          baseDelay: 1,
          retryIf: () => false,
        },
      ),
    ).rejects.toThrow("fatal");
    expect(attempt).toBe(1);
  });

  it("maxAttempts = 1 executes once with no retries", async () => {
    let attempt = 0;
    await expect(
      retry(
        () => {
          attempt++;
          return Promise.reject(new Error("fail"));
        },
        { maxAttempts: 1 },
      ),
    ).rejects.toThrow("fail");
    expect(attempt).toBe(1);
  });

  // --- Validation ---

  it("throws RangeError for maxAttempts < 1", async () => {
    await expect(
      retry(() => Promise.resolve(), { maxAttempts: 0 }),
    ).rejects.toThrow(RangeError);
  });

  it("throws RangeError for negative baseDelay", async () => {
    await expect(
      retry(() => Promise.resolve(), { baseDelay: -1 }),
    ).rejects.toThrow(RangeError);
  });

  it("throws RangeError for maxDelay < baseDelay", async () => {
    await expect(
      retry(() => Promise.resolve(), { baseDelay: 100, maxDelay: 50 }),
    ).rejects.toThrow(RangeError);
  });

  it("throws RangeError for multiplier < 1", async () => {
    await expect(
      retry(() => Promise.resolve(), { multiplier: 0.5 }),
    ).rejects.toThrow(RangeError);
  });

  it("throws RangeError for non-integer maxAttempts", async () => {
    await expect(
      retry(() => Promise.resolve(), { maxAttempts: 2.5 }),
    ).rejects.toThrow(RangeError);
  });

  it("accepts baseDelay = 0 (no delay between retries)", async () => {
    let attempt = 0;
    const result = await retry(
      () => {
        attempt++;
        if (attempt < 2) return Promise.reject(new Error("fail"));
        return Promise.resolve("ok");
      },
      { baseDelay: 0, maxDelay: 0 },
    );
    expect(result).toBe("ok");
  });
});

File: packages/core/flutter/test/retry_test.dart
import 'package:flutter_test/flutter_test.dart';
import '../lib/src/retry.dart';

void main() {
  group('retry', () {
    test('returns value on first success', () async {
      final result = await retry(() async => 42);
      expect(result, 42);
    });

    test('retries on failure and succeeds on subsequent attempt', () async {
      var attempt = 0;
      final result = await retry(
        () async {
          attempt++;
          if (attempt < 3) throw Exception('fail $attempt');
          return 'ok';
        },
        const RetryOptions(
          baseDelay: Duration(milliseconds: 10),
          maxDelay: Duration(milliseconds: 100),
        ),
      );
      expect(result, 'ok');
      expect(attempt, 3);
    });

    test('throws the last error after all attempts are exhausted', () async {
      var attempt = 0;
      await expectLater(
        retry(
          () async {
            attempt++;
            throw Exception('always fails');
          },
          const RetryOptions(
            maxAttempts: 2,
            baseDelay: Duration(milliseconds: 10),
          ),
        ),
        throwsA(isA<Exception>().having(
          (e) => e.toString(),
          'message',
          contains('always fails'),
        )),
      );
      expect(attempt, 2);
    });

    test('defaults to 3 maxAttempts', () async {
      var attempt = 0;
      await expectLater(
        retry(
          () async {
            attempt++;
            throw Exception('fail');
          },
          const RetryOptions(baseDelay: Duration(milliseconds: 10)),
        ),
        throwsA(isA<Exception>()),
      );
      expect(attempt, 3);
    });

    test('retryIf: retries when predicate returns true', () async {
      var attempt = 0;
      final result = await retry(
        () async {
          attempt++;
          if (attempt < 2) throw Exception('transient');
          return 'ok';
        },
        RetryOptions(
          baseDelay: const Duration(milliseconds: 10),
          retryIf: (e) => e.toString().contains('transient'),
        ),
      );
      expect(result, 'ok');
    });

    test('retryIf: rethrows immediately when predicate returns false',
        () async {
      var attempt = 0;
      await expectLater(
        retry(
          () async {
            attempt++;
            throw Exception('fatal');
          },
          RetryOptions(
            maxAttempts: 5,
            baseDelay: const Duration(milliseconds: 10),
            retryIf: (_) => false,
          ),
        ),
        throwsA(isA<Exception>().having(
          (e) => e.toString(),
          'message',
          contains('fatal'),
        )),
      );
      expect(attempt, 1);
    });

    test('maxAttempts = 1 executes once with no retries', () async {
      var attempt = 0;
      await expectLater(
        retry(
          () async {
            attempt++;
            throw Exception('fail');
          },
          const RetryOptions(maxAttempts: 1),
        ),
        throwsA(isA<Exception>()),
      );
      expect(attempt, 1);
    });

    test('accepts baseDelay = Duration.zero (no delay)', () async {
      var attempt = 0;
      final result = await retry(
        () async {
          attempt++;
          if (attempt < 2) throw Exception('fail');
          return 'ok';
        },
        const RetryOptions(
          baseDelay: Duration.zero,
          maxDelay: Duration.zero,
        ),
      );
      expect(result, 'ok');
    });

    // --- Validation ---

    test('throws ArgumentError for maxAttempts < 1', () async {
      await expectLater(
        retry(
          () async => 1,
          const RetryOptions(maxAttempts: 0),
        ),
        throwsArgumentError,
      );
    });

    test('throws ArgumentError for negative baseDelay', () async {
      await expectLater(
        retry(
          () async => 1,
          const RetryOptions(baseDelay: Duration(milliseconds: -1)),
        ),
        throwsArgumentError,
      );
    });

    test('throws ArgumentError for maxDelay < baseDelay', () async {
      await expectLater(
        retry(
          () async => 1,
          const RetryOptions(
            baseDelay: Duration(milliseconds: 100),
            maxDelay: Duration(milliseconds: 50),
          ),
        ),
        throwsArgumentError,
      );
    });

    test('throws ArgumentError for multiplier < 1', () async {
      await expectLater(
        retry(
          () async => 1,
          const RetryOptions(multiplier: 0.5),
        ),
        throwsArgumentError,
      );
    });

    test('preserves stack trace on final rethrow', () async {
      await expectLater(
        retry(
          () async {
            throw StateError('traced');
          },
          const RetryOptions(maxAttempts: 1),
        ),
        throwsA(isA<StateError>().having(
          (e) => e.message,
          'message',
          'traced',
        )),
      );
    });
  });
}
